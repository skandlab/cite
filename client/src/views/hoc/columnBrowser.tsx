import React, { useState } from "react";
import styled from "@emotion/styled";
import {
	Button,
	Card,
	Form,
	Icon,
	Item,
	Popup,
	SemanticShorthandContent,
	Statistic,
} from "semantic-ui-react";
import Interweave from "interweave";
import { ItemGroupLoader } from "../containers/itemGroupLoader";
import { ItemIsPresentType } from "../../utils/backendRequests";

export interface ColumnBrowserType {
	index: number;
	value: string;
	mute: boolean;
	isChecked: boolean;
	description?: string;
}

interface Props {
	filterIndex: number;
	title: string;
	popupContent: SemanticShorthandContent;
	options: ColumnBrowserType[];
	itemIsPresent: ItemIsPresentType["itemIsPresent"];
	filteredOptions: ColumnBrowserType[];
	loading: boolean;
	handleToggleCheckbox: (
		index: number,
		selectedOption: ColumnBrowserType,
		selectedIndex: number
	) => void;
	handleReset: (index: number) => void;
}

export const ColumnBrowser = ({
	filterIndex,
	title,
	popupContent,
	options,
	itemIsPresent,
	filteredOptions,
	loading,
	handleToggleCheckbox,
	handleReset,
}: Props) => {
	const [inputText, updateInputText] = useState("");

	let checkboxArray1: React.ReactElement[] = [];
	let checkboxArray2: React.ReactElement[] = [];
	let checkboxArray3: React.ReactElement[] = [];
	let arr1Index = 0;
	let arr23Index = 0;

	options.forEach((option) => {
		if (option.value.toLowerCase().startsWith(inputText.toLowerCase())) {
			if (option.isChecked) {
				checkboxArray1.push(
					<StyledCheckbox
						key={option.value}
						mute={false}
						option={option}
						optionIndex={arr1Index}
						filterIndex={filterIndex}
						handleToggleCheckbox={handleToggleCheckbox}
					/>
				);
				arr1Index++;
			} else {
				if (!itemIsPresent[option.value]) {
					checkboxArray2.push(
						<StyledCheckbox
							key={option.value}
							mute={false}
							option={option}
							optionIndex={arr23Index}
							filterIndex={filterIndex}
							handleToggleCheckbox={handleToggleCheckbox}
						/>
					);
				} else {
					checkboxArray3.push(
						<StyledCheckbox
							key={option.value}
							mute={true}
							option={option}
							optionIndex={arr23Index}
							filterIndex={filterIndex}
							handleToggleCheckbox={handleToggleCheckbox}
						/>
					);
				}
				arr23Index++;
			}
		}
	});

	return (
		<Card>
			<Card.Content>
				<StyledCardHeader>
					<span>
						<Icon name="filter" /> {title}
					</span>
					{popupContent && (
						<Popup
							size="small"
							inverted
							hoverable
							content={
								<Interweave content={popupContent as string} />
							}
							trigger={<Icon name="info circle" />}
						/>
					)}
				</StyledCardHeader>
				<Card.Description>
					<StyledForm>
						<Form.Input
							data-tour={filterIndex.toString() + "_filterInput"}
							data-testid={
								filterIndex.toString() + "_filterInput"
							}
							icon="search"
							iconPosition="left"
							value={inputText}
							onChange={(_, { value }) => updateInputText(value)}
						/>
					</StyledForm>

					<StyledItemGroup
						data-tour={filterIndex.toString() + "_checkbox-list"}
					>
						{loading ? (
							<ItemGroupLoader />
						) : (
							<div
								data-testid={
									filterIndex.toString() + "_checkbox-list"
								}
							>
								{checkboxArray1}
								{checkboxArray2}
								{checkboxArray3}
							</div>
						)}
					</StyledItemGroup>
				</Card.Description>
			</Card.Content>
			<Card.Content extra>
				<Button
					data-tour={filterIndex.toString() + "_resetButton"}
					data-testid={filterIndex.toString() + "_resetButton"}
					size="mini"
					labelPosition="right"
					icon
					disabled={filteredOptions.length === 0}
					onClick={() => {
						updateInputText("");
						handleReset(filterIndex);
					}}
				>
					<Icon name="undo" />
					Reset Filter
				</Button>
				<Statistic
					data-tour={filterIndex.toString() + "_statistic"}
					data-testid={filterIndex.toString() + "_statistic"}
					className="xs"
					label="filtered"
					floated="right"
					value={filteredOptions.length + " / " + options.length}
				/>
			</Card.Content>
		</Card>
	);
};

interface StyledCheckboxProps {
	filterIndex: number;
	mute: boolean;
	option: ColumnBrowserType;
	optionIndex: number;
	handleToggleCheckbox: (
		filterIndex: number,
		selectedOption: ColumnBrowserType,
		selectedIndex: number
	) => void;
}

const StyledCheckbox = ({
	filterIndex,
	mute,
	option,
	optionIndex,
	handleToggleCheckbox,
}: StyledCheckboxProps) => (
	<StyledItemWrapper mute={mute}>
		<Item>
			<Item.Content>
				<Item.Description>
					<div
						data-testid={option.value}
						className={
							option.isChecked
								? "ui checked checkbox"
								: "ui checkbox"
						}
						onClick={() =>
							handleToggleCheckbox(
								filterIndex,
								{ ...option, isChecked: !option.isChecked },
								optionIndex
							)
						}
					>
						<input
							className="hidden"
							type="checkbox"
							tabIndex={0}
							checked={option.isChecked}
							onChange={() => null} // discard
							name={option.value + "_checkbox"}
						/>
						<label htmlFor={option.value + "_checkbox"}>
							{option.value}
						</label>
						{option.description && (
							<small>{option.description}</small>
						)}
					</div>
				</Item.Description>
			</Item.Content>
		</Item>
	</StyledItemWrapper>
);

const StyledItemWrapper = styled.div<{ mute: boolean }>`
	padding: 0.7em 0.1em;
	& label,
	small {
		color: ${(props) => (props.mute ? "#b3b3b3" : "#4a4f59")} !important;
	}
`;

const StyledItemGroup = styled(Item.Group)`
	overflow-y: scroll;
	height: 224px;
	min-height: 224px;
	margin-top: 0 !important;
`;

const StyledCardHeader = styled(Card.Meta)`
	display: flex;
	justify-content: space-between;
`;

const StyledForm = styled(Form)`
	& .field:last-child {
		margin-bottom: 0.5em !important;
	}
`;
