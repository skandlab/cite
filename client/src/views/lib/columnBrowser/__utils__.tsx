import { screen } from "@testing-library/dom";

export function TestState(
	filterValue: string,
	checkboxList: { item: string; checked: boolean }[],
	statistic: string,
	disabled: boolean
) {
	expect(getFilterInput()).toHaveValue(filterValue);
	checkboxList.forEach(({ item, checked }) => {
		checked
			? expect(getCheckbox(item)).toHaveClass("ui checked checkbox")
			: expect(getCheckbox(item)).not.toHaveClass("ui checked checkbox");
	});
	expect(getCheckboxListChilds()).toHaveLength(checkboxList.length);
	expect(getStatistic()).toHaveTextContent(statistic);
	disabled
		? expect(getResetButton()).toBeDisabled()
		: expect(getResetButton()).not.toBeDisabled();
}

export function getFilterInput() {
	return screen.getByTestId("filterInput").getElementsByTagName("input")[0];
}

export function getCheckboxListChilds() {
	return screen.getByTestId("checkbox-list").childNodes;
}

export function getCheckboxToggle(item: string) {
	return getCheckbox(item).getElementsByTagName("input")[0];
}

export function getCheckbox(item: string) {
	return screen.getByTestId(item);
}

export function getResetButton() {
	return screen.getByTestId("resetButton");
}

export function getStatistic() {
	return screen.getByTestId("statistic");
}
