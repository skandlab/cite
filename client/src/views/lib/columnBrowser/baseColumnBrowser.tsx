import React, { useState } from "react";
import styled from "@emotion/styled";
import {
	Button,
	Card,
	Checkbox,
	Form,
	Icon,
	Item,
	Popup,
	SemanticShorthandContent,
	Statistic,
} from "semantic-ui-react";

export interface ColumnBrowserType {
	index: number;
	value: string;
	mute: boolean;
	isChecked: boolean;
	description?: string;
}

interface Props {
	title: string;
	options: ColumnBrowserType[];
	totalChecked: number;
	popupContent: SemanticShorthandContent;
	handleToggleCheckbox: (option: ColumnBrowserType, index: number) => void;
	handleReset: () => void;
}

/**
 *
 * @param props
 *
 * to test:
 * rendering:
 * 		1. title
 * 		2. checkboxlist
 * 		3, reset button
 * 		4. no of checked
 * events:
 * 		1. input filter
 */
export const BaseColumnBrowser = (props: Props) => {
	const [inputText, updateInputText] = useState("");

	return (
		<Card>
			<Card.Content>
				<StyledCardHeader>
					<span>
						<Icon name="filter" /> {props.title}
					</span>
					{props.popupContent && (
						<Popup
							size="small"
							inverted
							hoverable
							content={props.popupContent}
							trigger={<Icon name="info circle" />}
						/>
					)}
				</StyledCardHeader>
				<Card.Description>
					<StyledForm>
						<Form.Input
							data-testid="filterInput"
							icon="search"
							iconPosition="left"
							value={inputText}
							onChange={(_, { value }) => updateInputText(value)}
						/>
					</StyledForm>

					<StyledItemGroup data-testid="checkbox-list">
						{props.options.map((option, index) =>
							option.value
								.toLowerCase()
								.startsWith(inputText.toLowerCase()) ? (
								<StyledItem
									key={option.value}
									mute={option.mute ? "" : undefined}
								>
									<Item.Content>
										<Item.Description>
											<Checkbox
												data-testid={option.value}
												label={option.value}
												checked={option.isChecked}
												onChange={() =>
													props.handleToggleCheckbox(
														option,
														index
													)
												}
											/>
										</Item.Description>
										{option.description && (
											<StyledItemExtra key={option.value}>
												<small>
													{option.description}
												</small>
											</StyledItemExtra>
										)}
									</Item.Content>
								</StyledItem>
							) : null
						)}
					</StyledItemGroup>
				</Card.Description>
			</Card.Content>
			<Card.Content extra>
				<Button
					data-testid="resetButton"
					size="mini"
					labelPosition="right"
					icon
					disabled={props.totalChecked === 0}
					onClick={() => {
						updateInputText("");
						props.handleReset();
					}}
				>
					<Icon name="undo" />
					Reset Filter
				</Button>
				<Statistic
					data-testid="statistic"
					className="xs"
					label="showing"
					floated="right"
					value={props.totalChecked + " / " + props.options.length}
				/>
			</Card.Content>
		</Card>
	);
};

const StyledItem = styled(Item)`
	margin: 0 !important;
	padding: 0 !important;

	& label {
		color: ${(props) =>
			props.mute !== undefined ? "#b3b3b3" : "#4a4f59"} !important;
	}
`;

const StyledItemGroup = styled(Item.Group)`
	overflow-y: scroll;
	height: 224px;
	min-height: 224px;
	margin-top: 0 !important;
`;

const StyledCardHeader = styled(Card.Meta)`
	display: flex;
	justify-content: space-between;
`;

const StyledForm = styled(Form)`
	& .field:last-child {
		margin-bottom: 0.5em !important;
	}
`;

const StyledItemExtra = styled(Item.Extra)`
	margin: 0 !important;
	color: rgba(0, 0, 0, 0.6) !important;
`;
