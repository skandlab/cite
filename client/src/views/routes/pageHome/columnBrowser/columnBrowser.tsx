import React, { useState } from "react";
import styled from "@emotion/styled";

import {
	Item,
	Checkbox,
	Icon,
	Form,
	Card,
	Statistic,
	Button,
	Dimmer,
	Loader,
	Popup,
	SemanticShorthandContent,
} from "semantic-ui-react";

import { InterfaceColumnBrowserItem } from "../../../../utils/interfaces";

export interface InterfaceFilteredOptionsProps
	extends InterfaceColumnBrowserItem {
	orginalIndex: number;
}

interface Props {
	title: string;
	filteredOptions: InterfaceFilteredOptionsProps[];
	options: InterfaceColumnBrowserItem[];
	handleFilter: (
		filteredOptions: InterfaceFilteredOptionsProps[],
		options: InterfaceColumnBrowserItem[]
	) => void;
	isFetchingData: boolean;
	popupContent: SemanticShorthandContent;
}

const StyledCardHeader = styled(Card.Meta)`
	display: flex;
	justify-content: space-between;
`;

const StyledFormGroup = styled(Form.Group)`
	margin-bottom: 0 !important;
`;

const StyledItemGroup = styled(Item.Group)`
	overflow-y: scroll;
	height: 224px;
	min-height: 224px;
	margin-top: 0 !important;
`;

const StyledItem = styled(Item)<{ mute: string | undefined }>`
	& label {
		color: ${(props) =>
			props.mute !== undefined ? "#b3b3b3" : "#4a4f59"} !important;
	}
`;

const StyledItemExtra = styled(Item.Extra)`
	margin-bottom: 16px;
	line-height: 1em;
`;

const CheckboxItem = ({
	index,
	option,
	handleToggleCheckbox,
}: {
	index: number;
	option: InterfaceColumnBrowserItem;
	handleToggleCheckbox: Function;
}) => (
	<StyledItem mute={option.mute ? "" : undefined}>
		<Item.Content>
			<Item.Description>
				<Checkbox
					label={option.value}
					checked={option.isChecked}
					onChange={() =>
						handleToggleCheckbox(index, option.isChecked)
					}
				/>
			</Item.Description>
			{option.description && (
				<StyledItemExtra key={option.value}>
					<small>{option.description}</small>
				</StyledItemExtra>
			)}
		</Item.Content>
	</StyledItem>
);

export const ColumnBrowser = ({
	title,
	filteredOptions,
	options,
	handleFilter,
	isFetchingData,
	popupContent,
}: Props) => {
	const [inputText, updateInputText] = useState("");

	const handleToggleCheckbox = (index: number, prevChecked: boolean) => {
		let tmpFilteredOptions: InterfaceFilteredOptionsProps[] = prevChecked
			? [
					...filteredOptions.slice(0, index),
					...filteredOptions.slice(index + 1, filteredOptions.length),
			  ]
			: [
					...filteredOptions,
					{
						...options[index],
						isChecked: !prevChecked,
						orginalIndex: index,
					},
			  ];
		let currentIndex = prevChecked
			? filteredOptions[index].orginalIndex
			: index;
		let tmpOptions: InterfaceColumnBrowserItem[] = [
			...options.slice(0, currentIndex),
			{ ...options[currentIndex], isChecked: !prevChecked },
			...options.slice(currentIndex + 1, options.length),
		];
		handleFilter(tmpFilteredOptions, tmpOptions);
	};

	return (
		<Card>
			<Card.Content>
				<StyledCardHeader>
					<span>
						<Icon name="filter" /> {title}
					</span>
					{popupContent && (
						<Popup
							size="small"
							inverted
							hoverable
							content={popupContent}
							trigger={<Icon name="info circle" />}
						/>
					)}
				</StyledCardHeader>
				<Card.Description>
					<Form>
						<StyledFormGroup widths="equal">
							<Form.Input
								icon="search"
								iconPosition="left"
								value={inputText}
								onChange={(_, { value }) =>
									updateInputText(value)
								}
							/>
						</StyledFormGroup>
					</Form>

					<StyledItemGroup>
						{!isFetchingData ? (
							<div>
								{filteredOptions
									.filter((option) =>
										option.value
											.toLowerCase()
											.startsWith(inputText.toLowerCase())
									)
									.map(
										(
											option: InterfaceFilteredOptionsProps,
											index: number
										) => (
											<CheckboxItem
												key={option.value}
												index={index}
												option={option}
												handleToggleCheckbox={
													handleToggleCheckbox
												}
											/>
										)
									)}
								{options
									.filter((option) =>
										option.value
											.toLowerCase()
											.startsWith(inputText.toLowerCase())
									)
									.map(
										(option) =>
											!option.isChecked && (
												<CheckboxItem
													key={option.value}
													index={option.index}
													option={option}
													handleToggleCheckbox={
														handleToggleCheckbox
													}
												/>
											)
									)}
							</div>
						) : (
							<Dimmer active inverted>
								<Loader />
							</Dimmer>
						)}
					</StyledItemGroup>
				</Card.Description>
			</Card.Content>
			<Card.Content extra>
				<Button
					size="mini"
					icon
					disabled={filteredOptions.length === 0}
					labelPosition="right"
					onClick={() => {
						// reset filter
						let tmp = options;
						filteredOptions.forEach(
							(option) =>
								(tmp[option.orginalIndex].isChecked = false)
						);
						handleFilter([], tmp);
						updateInputText("");
					}}
				>
					<Icon name="undo" />
					Reset Filter
				</Button>
				<Statistic
					className="xs"
					label="showing"
					value={
						(filteredOptions.length === 0
							? options.length
							: filteredOptions.length) +
						" / " +
						options.length
					}
					floated="right"
				/>
			</Card.Content>
		</Card>
	);
};
