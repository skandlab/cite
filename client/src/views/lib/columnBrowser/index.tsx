import { ColumnBrowserType as interfaceType } from "./baseColumnBrowser";

export * from "./statefulColumnBrowser";
export * from "./statelessColumnBrowser";
export type ColumnBrowserType = interfaceType;
