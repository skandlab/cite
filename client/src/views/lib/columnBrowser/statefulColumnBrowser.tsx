import React, { useReducer, useEffect } from "react";
import { SemanticShorthandContent } from "semantic-ui-react";
import { BaseColumnBrowser, ColumnBrowserType } from "./baseColumnBrowser";

interface ColumnBrowserProps {
	title: string;
	options: ColumnBrowserType[];
	popupContent: SemanticShorthandContent;
	handleToggleCheckbox: (changedOptions: ColumnBrowserType[]) => void;
}

interface State {
	options: ColumnBrowserType[];
	totalChecked: number;
}

const initialState: State = {
	options: [],
	totalChecked: 0,
};

function reducer(
	state: State,
	action: {
		type: string;
		payload: any;
	}
) {
	switch (action.type) {
		case "UPDATE_INTIAL_OPTIONS":
			return {
				...state,
				options: action.payload as ColumnBrowserType[],
				totalChecked: (action.payload as ColumnBrowserType[]).filter(
					(option) => option.isChecked
				).length,
			};
		case "RESET_STATE":
			let resetStateOptions = (action.payload
				.options as ColumnBrowserType[])
				.map((option) => ({ ...option, isChecked: false }))
				.sort((a, b) => a.index - b.index);
			action.payload.handleToggleCheckbox(resetStateOptions.slice(0, 0));
			return {
				...initialState,
				options: resetStateOptions,
				totalChecked: 0,
			};
		case "TOGGLE_CHECKBOX":
			let { option, index, handleToggleCheckbox } = action.payload;
			let tmpOptions = [];
			let totalChecked = 0;

			if (option.isChecked) {
				let futureTotalChecked = state.totalChecked - 1;
				tmpOptions = [
					...state.options.slice(0, index),
					...state.options.slice(index + 1),
					{ ...state.options[index], isChecked: false },
				];
				tmpOptions = [
					...tmpOptions.slice(0, futureTotalChecked),
					...tmpOptions
						.slice(futureTotalChecked)
						.sort((a, b) => a.index - b.index),
				];
				totalChecked = futureTotalChecked;
			} else {
				tmpOptions = [
					{ ...state.options[index], isChecked: true },
					...state.options.slice(0, index),
					...state.options.slice(index + 1),
				];
				totalChecked = state.totalChecked + 1;
			}
			handleToggleCheckbox(tmpOptions.slice(0, totalChecked));
			return {
				...state,
				options: tmpOptions,
				totalChecked: totalChecked,
			};
		default:
			return state;
	}
}

export const StatefulColumnBrowser = (props: ColumnBrowserProps) => {
	const { options, handleToggleCheckbox } = props;
	const [state, dispatch] = useReducer(reducer, initialState);

	useEffect(
		() => dispatch({ type: "UPDATE_INTIAL_OPTIONS", payload: options }),
		[options]
	);

	return (
		<BaseColumnBrowser
			{...props}
			{...state}
			handleToggleCheckbox={(option, index) =>
				dispatch({
					type: "TOGGLE_CHECKBOX",
					payload: {
						option: option,
						index: index,
						handleToggleCheckbox: handleToggleCheckbox,
					},
				})
			}
			handleReset={() =>
				dispatch({
					type: "RESET_STATE",
					payload: {
						options: options,
						handleToggleCheckbox: handleToggleCheckbox,
					},
				})
			}
		/>
	);
};
