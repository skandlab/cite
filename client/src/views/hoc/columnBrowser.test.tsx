import React from "react";
import "@testing-library/dom";
import "@testing-library/jest-dom/extend-expect";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import { ColumnBrowser } from "./columnBrowser";

const mockedHandleToggleCheckbox = jest.fn((data) => null);
const mockedHandleReset = jest.fn((data) => null);

const options = [
	{
		index: 0,
		value: "cc",
		mute: false,
		isChecked: false,
	},
	{
		index: 1,
		value: "cs",
		mute: false,
		isChecked: false,
	},
	{
		index: 2,
		value: "nn",
		mute: false,
		isChecked: false,
	},
	{
		index: 3,
		value: "sc",
		mute: false,
		isChecked: false,
	},
];

/**
 * nothing is selected
 */
const NoCheckedRender = () =>
	render(
		<ColumnBrowser
			index={0}
			title="Interaction type"
			popupContent={
				<p>if no associated heatmap present then gene will be dimmed</p>
			}
			options={options}
			loading={false}
			filteredOptions={[]}
			handleToggleCheckbox={mockedHandleToggleCheckbox}
			handleReset={mockedHandleReset}
		/>
	);

/**
 * one item is checked
 */
const OneCheckedRender = () =>
	render(
		<ColumnBrowser
			index={0}
			title="Interaction type"
			popupContent={
				<p>if no associated heatmap present then gene will be dimmed</p>
			}
			options={[
				...options.slice(0, 3),
				{ ...options[3], isChecked: true },
			]}
			loading={false}
			filteredOptions={[{ ...options[3], isChecked: true }]}
			handleToggleCheckbox={mockedHandleToggleCheckbox}
			handleReset={mockedHandleReset}
		/>
	);

afterEach(() => {
	cleanup();
	mockedHandleToggleCheckbox.mockClear();
	mockedHandleReset.mockClear();
});

describe("snapshot testing", () => {
	test("initial render", () => {
		const { asFragment } = NoCheckedRender();
		expect(asFragment()).toMatchSnapshot();

		TestState(
			"",
			[
				{ item: "cc", checked: false },
				{ item: "cs", checked: false },
				{ item: "nn", checked: false },
				{ item: "sc", checked: false },
			],
			"0 / 4",
			true
		);
	});

	test("one item is checked", () => {
		const { asFragment } = OneCheckedRender();
		expect(asFragment()).toMatchSnapshot();

		TestState(
			"",
			[
				{ item: "sc", checked: true },
				{ item: "cc", checked: false },
				{ item: "cs", checked: false },
				{ item: "nn", checked: false },
			],
			"1 / 4",
			false
		);
	});
});

describe("test method calls", () => {
	test("toggle", () => {
		NoCheckedRender();

		/**
		 * toggle nn
		 */
		fireEvent.click(getCheckboxToggle("nn"));
		expect(mockedHandleToggleCheckbox.mock.calls[0]).toEqual([
			0,
			{ ...options[2], isChecked: true },
			2,
		]);
	});

	test("untoggle", () => {
		OneCheckedRender();

		/**
		 * toggle sc
		 */
		fireEvent.click(getCheckboxToggle("sc"));
		expect(mockedHandleToggleCheckbox.mock.calls[0]).toEqual([
			0,
			{ ...options[3], isChecked: false },
			0,
		]);
	});

	test("reset", () => {
		OneCheckedRender();

		/**
		 * reset
		 */
		fireEvent.click(getResetButton());
		expect(mockedHandleReset.mock.calls[0][0]).toEqual(0);
	});
});

describe("test filter input", () => {
	test("filter n -> filter c -> reset", () => {
		OneCheckedRender();

		/**
		 * filter "n"
		 */
		fireEvent.change(getFilterInput(), { target: { value: "n" } });
		TestState("n", [{ item: "nn", checked: false }], "1 / 4", false);

		fireEvent.change(getFilterInput(), { target: { value: "s" } });
		TestState("s", [{ item: "sc", checked: true }], "1 / 4", false);

		/**
		 * reset,
		 * untoggling does not take place
		 * but check that input field is cleared
		 */
		fireEvent.click(getResetButton());
		expect(mockedHandleReset.mock.calls[0][0]).toEqual(0);
		TestState(
			"",
			[
				{ item: "cc", checked: false },
				{ item: "cs", checked: false },
				{ item: "nn", checked: false },
				{ item: "sc", checked: true },
			],
			"1 / 4",
			false
		);
	});

	test("filter -> clear", () => {
		NoCheckedRender();

		/**
		 * input change
		 */
		fireEvent.change(getFilterInput(), { target: { value: "c" } });
		TestState(
			"c",
			[
				{ item: "cc", checked: false },
				{ item: "cs", checked: false },
			],
			"0 / 4",
			true
		);

		/**
		 * clear input
		 */
		fireEvent.change(getFilterInput(), { target: { value: "" } });
		TestState(
			"",
			[
				{ item: "cc", checked: false },
				{ item: "cs", checked: false },
				{ item: "nn", checked: false },
				{ item: "sc", checked: false },
			],
			"0 / 4",
			true
		);
	});
});

/**
 * test ignored because disabled button is registering clicks
 * see the following links:
 *  https://github.com/testing-library/dom-testing-library/issues/92
 *  https://github.com/whatwg/html/issues/2368
 */
// test("filter -> reset", () => {});

/**
 * utils
 */
export function TestState(
	filterValue: string,
	checkboxList: { item: string; checked: boolean }[],
	statistic: string,
	disabled: boolean
) {
	expect(getFilterInput()).toHaveValue(filterValue);
	checkboxList.forEach(({ item, checked }) => {
		checked
			? expect(getCheckbox(item)).toHaveClass("ui checked checkbox")
			: expect(getCheckbox(item)).not.toHaveClass("ui checked checkbox");
	});
	expect(getCheckboxListChilds()).toHaveLength(checkboxList.length);
	expect(getStatistic()).toHaveTextContent(statistic);
	disabled
		? expect(getResetButton()).toBeDisabled()
		: expect(getResetButton()).not.toBeDisabled();
}

export function getFilterInput() {
	return screen.getByTestId("0_filterInput").getElementsByTagName("input")[0];
}

export function getCheckboxListChilds() {
	return screen.getByTestId("0_checkbox-list").childNodes;
}

export function getCheckboxToggle(item: string) {
	return getCheckbox(item).getElementsByTagName("input")[0];
}

export function getCheckbox(item: string) {
	return screen.getByTestId(item);
}

export function getResetButton() {
	return screen.getByTestId("0_resetButton");
}

export function getStatistic() {
	return screen.getByTestId("0_statistic");
}
