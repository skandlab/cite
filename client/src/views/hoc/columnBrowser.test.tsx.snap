// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing initial render 1`] = `
<DocumentFragment>
  <div
    class="ui card"
  >
    <div
      class="content"
    >
      <div
        class="meta css-106mbhu"
      >
        <span>
          <i
            aria-hidden="true"
            class="filter icon"
          />
           Interaction type
        </span>
        <i
          aria-hidden="true"
          class="info circle icon"
        />
      </div>
      <div
        class="description"
      >
        <form
          class="ui form css-1rrhbx0"
        >
          <div
            class="field"
          >
            <div
              class="ui left icon input"
              data-testid="0_filterInput"
            >
              <input
                type="text"
                value=""
              />
              <i
                aria-hidden="true"
                class="search icon"
              />
            </div>
          </div>
        </form>
        <div
          class="ui items css-p5nlxd"
        >
          <div
            data-testid="0_checkbox-list"
          >
            <div
              class="css-tescqt"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="description"
                  >
                    <div
                      class="ui checkbox"
                      data-testid="cc"
                    >
                      <input
                        class="hidden"
                        name="cc_checkbox"
                        tabindex="0"
                        type="checkbox"
                      />
                      <label
                        for="cc_checkbox"
                      >
                        cc
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-tescqt"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="description"
                  >
                    <div
                      class="ui checkbox"
                      data-testid="cs"
                    >
                      <input
                        class="hidden"
                        name="cs_checkbox"
                        tabindex="0"
                        type="checkbox"
                      />
                      <label
                        for="cs_checkbox"
                      >
                        cs
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-tescqt"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="description"
                  >
                    <div
                      class="ui checkbox"
                      data-testid="nn"
                    >
                      <input
                        class="hidden"
                        name="nn_checkbox"
                        tabindex="0"
                        type="checkbox"
                      />
                      <label
                        for="nn_checkbox"
                      >
                        nn
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-tescqt"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="description"
                  >
                    <div
                      class="ui checkbox"
                      data-testid="sc"
                    >
                      <input
                        class="hidden"
                        name="sc_checkbox"
                        tabindex="0"
                        type="checkbox"
                      />
                      <label
                        for="sc_checkbox"
                      >
                        sc
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="extra content"
    >
      <button
        class="ui mini icon disabled right labeled button"
        data-testid="0_resetButton"
        disabled=""
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="undo icon"
        />
        Reset Filter
      </button>
      <div
        class="ui right floated statistic xs"
        data-testid="0_statistic"
      >
        <div
          class="value"
        >
          0 / 4
        </div>
        <div
          class="label"
        >
          showing
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`snapshot testing one item is checked 1`] = `
<DocumentFragment>
  <div
    class="ui card"
  >
    <div
      class="content"
    >
      <div
        class="meta css-106mbhu"
      >
        <span>
          <i
            aria-hidden="true"
            class="filter icon"
          />
           Interaction type
        </span>
        <i
          aria-hidden="true"
          class="info circle icon"
        />
      </div>
      <div
        class="description"
      >
        <form
          class="ui form css-1rrhbx0"
        >
          <div
            class="field"
          >
            <div
              class="ui left icon input"
              data-testid="0_filterInput"
            >
              <input
                type="text"
                value=""
              />
              <i
                aria-hidden="true"
                class="search icon"
              />
            </div>
          </div>
        </form>
        <div
          class="ui items css-p5nlxd"
        >
          <div
            data-testid="0_checkbox-list"
          >
            <div
              class="css-tescqt"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="description"
                  >
                    <div
                      class="ui checked checkbox"
                      data-testid="sc"
                    >
                      <input
                        checked=""
                        class="hidden"
                        name="sc_checkbox"
                        tabindex="0"
                        type="checkbox"
                      />
                      <label
                        for="sc_checkbox"
                      >
                        sc
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-tescqt"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="description"
                  >
                    <div
                      class="ui checkbox"
                      data-testid="cc"
                    >
                      <input
                        class="hidden"
                        name="cc_checkbox"
                        tabindex="0"
                        type="checkbox"
                      />
                      <label
                        for="cc_checkbox"
                      >
                        cc
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-tescqt"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="description"
                  >
                    <div
                      class="ui checkbox"
                      data-testid="cs"
                    >
                      <input
                        class="hidden"
                        name="cs_checkbox"
                        tabindex="0"
                        type="checkbox"
                      />
                      <label
                        for="cs_checkbox"
                      >
                        cs
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-tescqt"
            >
              <div
                class="item"
              >
                <div
                  class="content"
                >
                  <div
                    class="description"
                  >
                    <div
                      class="ui checkbox"
                      data-testid="nn"
                    >
                      <input
                        class="hidden"
                        name="nn_checkbox"
                        tabindex="0"
                        type="checkbox"
                      />
                      <label
                        for="nn_checkbox"
                      >
                        nn
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="extra content"
    >
      <button
        class="ui mini icon right labeled button"
        data-testid="0_resetButton"
      >
        <i
          aria-hidden="true"
          class="undo icon"
        />
        Reset Filter
      </button>
      <div
        class="ui right floated statistic xs"
        data-testid="0_statistic"
      >
        <div
          class="value"
        >
          1 / 4
        </div>
        <div
          class="label"
        >
          showing
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
